migration/migration.go

```diff
@@ -1,25 +1,136 @@
+			return -1, err
+		}
+	}
+
+	return current, nil
+}
+
+// fn defines the type of function that can be part of a migration sequence
+type fn func(tx *gorm.DB) error
+
+// executeSQLFile loads the given filename from the packaged SQL files and
+// executes it on the given database
+func executeSQLFile(filename string) fn {
+	return func(db *gorm.DB) error {
+		data, err := Asset(filename)
+		if err != nil {
+			return err
+		}
+		return db.Exec(string(data)).Error
+	}
+}
+
+// Migrate executes the required migration of the database on startup
+func Migrate(db *gorm.DB) error {
+
+	migrations := [][]fn{}
+
+	// Version 0
+	migrations = append(migrations, []fn{executeSQLFile("000-bootstrap.sql")})
+
+	// Version 1
+	migrations = append(migrations, []fn{executeSQLFile("001-common.sql")})
+
+	// Version N
+	//
+	// In order to add an upgrade, simply append an array of MigrationFunc to the
+	// the end of the "migrations" slice. The version numbers are determined by
+	// the index in the array. The following code in comments show how you can
+	// do a migration in 3 steps. If one of the steps fails, the others are not
+	// executed.
+	// If something goes wrong during the migration, all you need to do is return
+	// an error that is not nil.
+
+	/*
+		migrations = append(migrations, []Func{
+			func(db *gorm.DB) error {
+				// Execute random go code
+				return nil
+			},
+			executeSQLFile("YOUR_OWN_FILE.sql"),
+			func(db *gorm.DB) error {
+				// Execute random go code
+				return nil
+			},
+		})
+	*/
+
+	ts := models.NewGormTransactionSupport(db)
+	var err error
+	for nextVersion := int64(0); nextVersion < int64(len(migrations)) && err == nil; nextVersion++ {
+
+		err = transaction.Do(ts, func() error {
+			// Determine current version and adjust the outmost loop
+			// iterator variable "version"
+			currentVersion, err := getCurrentVersion(ts.TX())
+			if err != nil {
+				return err
+			}
+			nextVersion = currentVersion + 1
+			if nextVersion >= int64(len(migrations)) {
+				// No further updates to apply (this is NOT an error)
+				return nil
+			}
+
+			log.Printf("Attempt to update DB to version %d\n", nextVersion)
+
+			// Apply all the updates of the next version
+			for j := range migrations[nextVersion] {
+				if err = migrations[nextVersion][j](ts.TX().Debug()); err != nil {
+					return err
+				}
+			}
+
+			// Finalize the update by inserting the new version into the version table
+			err = ts.TX().Debug().Exec("insert into version (version) values(?)", nextVersion).Error
+			if err != nil {
+				log.Printf("Failed to update DB to version %d: %s\n", nextVersion, err)
+			} else {
+				log.Printf("Successfully updated DB to version %d\n", nextVersion)
+			}
+			return err
+		})
+
 	}
 
+	return err
+}
+
+// Populate makes sure the database is populated with the correct types (e.g. system.bug etc.)
+func Populate(ctx context.Context, db *gorm.DB, witr models.WorkItemTypeRepository) error {
```

Comment:
I understand that this method is called conditionally based on the configuration flag for populate.commontypes.

Could we rename the method to PopulateCommonTypes , or in a more general way PopulateSeededData ( considering everything that goes in here is for ALM Online only ) ?

I just felt that the name of the method wasn't very reflective of what it's really doing. Unless we plan to generalize the body of the method, this should be named differently.